### PR TITLE
feat(contracts): bridge costs optimization

### PR DESCRIPTION
**Pull request by zimpha**
### Purpose or design rationale of this PR

This PR try to reduce the gas costs for bridging. The following changes are made:

- change `counterpart`, `router`, `messenger` in gateway from storage variable to immutable variable
- change `rollup` and `messageQueue` in `L1ScrollMessenger` from storage variable to immutable variable
- change `messageQueue` in `ScrollChain` from storage variable to immutable variable
- change `messenger` and `scrollChain` in `L1MessageQueue` from storage variable to immutable variable
- add `L1MessageQueueWithGasPriceOracle` to merge `L2GasPriceOracle` and `L1MessageQueue`

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [x] perf: A code change that improves performance
- [x] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [x] No, this PR doesn't involve a new deployment, git tag, docker image tag

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change

### CODE REVIEW

File: contracts/src/L1/rollup/L1MessageQueue.sol [hunk 1]
```diff
@@ -20,34 +20,25 @@ import {AddressAliasHelper} from "../../libraries/common/AddressAliasHelper.sol"
...
+     *************/
+
+    /// @dev The storage slot used as L1ScrollMessenger contract, which is deprecated now.
+    address private __messenger;
+
+    /// @dev The storage slot used as ScrollChain contract, which is deprecated now.
+    address private __scrollChain;
 
     /// @notice The address EnforcedTxGateway contract.
     address public enforcedTxGateway;
```
Comment (Thegaram) [inline contracts/src/L1/rollup/L1MessageQueue.sol]: Should we make `enforcedTxGateway` and `gasOracle` immutable too?

File: contracts/src/L1/rollup/ScrollChain.sol [hunk 2]
```diff
@@ -48,15 +48,18 @@ contract ScrollChain is OwnableUpgradeable, PausableUpgradeable, IScrollChain {
...
     /// @notice The maximum number of transactions allowed in each chunk.
     uint256 public maxNumTxInChunk;
 
-    /// @notice The address of L1MessageQueue.
-    address public messageQueue;
+    /// @dev The storage slot used as L1MessageQueue contract, which is deprecated now.
+    address private __messageQueue;
 
     /// @notice The address of RollupVerifier.
     address public verifier;
```
Comment (Thegaram) [inline contracts/src/L1/rollup/ScrollChain.sol]: Should we also make `verifier` immutable? This won't change often.

Comment (zimpha) [reply contracts/src/L1/rollup/ScrollChain.sol]: Yeah, we can.

File: contracts/src/L1/rollup/L1MessageQueue.sol (hunk 1 above)
Comment (zimpha) [reply contracts/src/L1/rollup/L1MessageQueue.sol]: `gasOracle` maybe changed often, also we won't use `L1MessageQueue` as the implementation of `L1MessageQueue` proxy.

`enforcedTxGateway` can be changed to immutable.

**Comment by zimpha** [comment]
change `it.skip` to `it` in `integration-test/GasOptimizationUpgrade.spec.ts` and then run the following command.

```bash
npx hardhat test integration-test/GasOptimizationUpgrade.spec.ts
```

below are example results:

```
  GasOptimizationUpgrade.spec
    L1 upgrade
L1ETHGateway.depositETH before upgrade: GasUsed[271677]
L1GatewayRouter.depositETH before upgrade: GasUsed[295114]
L1ScrollMessenger.sendMessage before upgrade: GasUsed[195609]
L1ETHGateway.depositETH after upgrade: GasUsed[161598]
L1GatewayRouter.depositETH after upgrade: GasUsed[185035]
L1ScrollMessenger.sendMessage after upgrade: GasUsed[133910]
      ✔ should succeed on L1ETHGateway (23503ms)
L1WETHGateway.depositERC20 WETH before upgrade: GasUsed[317736]
L1GatewayRouter.depositERC20 WETH before upgrade: GasUsed[343644]
L1WETHGateway.depositERC20 WETH before upgrade: GasUsed[193254]
L1GatewayRouter.depositERC20 WETH before upgrade: GasUsed[219162]
      ✔ should succeed on L1WETHGateway (7843ms)
L1StandardERC20Gateway.depositERC20 USDT before upgrade: GasUsed[347763]
L1GatewayRouter.depositERC20 USDT before upgrade: GasUsed[375859]
L1StandardERC20Gateway.depositERC20 USDT before upgrade: GasUsed[203309]
L1GatewayRouter.depositERC20 USDT before upgrade: GasUsed[231405]
      ✔ should succeed on L1StandardERC20Gateway (6706ms)
L1CustomERC20Gateway.depositERC20 DAI before upgrade: GasUsed[313758]
L1GatewayRouter.depositERC20 DAI before upgrade: GasUsed[339666]
L1CustomERC20Gateway.depositERC20 DAI before upgrade: GasUsed[188412]
L1GatewayRouter.depositERC20 DAI before upgrade: GasUsed[214320]
      ✔ should succeed on L1CustomERC20Gateway (5778ms)
L1USDCGateway.depositERC20 USDC before upgrade: GasUsed[337606]
L1GatewayRouter.depositERC20 USDC before upgrade: GasUsed[365463]
L1USDCGateway.depositERC20 USDC before upgrade: GasUsed[212482]
L1GatewayRouter.depositERC20 USDC before upgrade: GasUsed[240339]
      ✔ should succeed on L1USDCGateway (8311ms)
    L2 upgrade
L2ETHGateway.withdrawETH before upgrade: GasUsed[168104]
L2GatewayRouter.withdrawETH before upgrade: GasUsed[191358]
L2ScrollMessenger.sendMessage before upgrade: GasUsed[133257]
L2ETHGateway.withdrawETH after upgrade: GasUsed[166831]
L2GatewayRouter.withdrawETH after upgrade: GasUsed[183040]
L2ScrollMessenger.sendMessage after upgrade: GasUsed[138302]
      ✔ should succeed on L2ETHGateway (284324ms)
L2WETHGateway.withdrawERC20 WETH before upgrade: GasUsed[203795]
L2GatewayRouter.withdrawERC20 WETH before upgrade: GasUsed[220681]
L2WETHGateway.withdrawERC20 WETH before upgrade: GasUsed[194408]
L2GatewayRouter.withdrawERC20 WETH before upgrade: GasUsed[217384]
      ✔ should succeed on L2WETHGateway (20490ms)
L2StandardERC20Gateway.withdrawERC20 USDT before upgrade: GasUsed[185916]
L2GatewayRouter.withdrawERC20 USDT before upgrade: GasUsed[204972]
L2StandardERC20Gateway.withdrawERC20 USDT before upgrade: GasUsed[176553]
L2GatewayRouter.withdrawERC20 USDT before upgrade: GasUsed[201699]
      ✔ should succeed on L2StandardERC20Gateway (22963ms)
L2CustomERC20Gateway.withdrawERC20 DAI before upgrade: GasUsed[190416]
L2GatewayRouter.withdrawERC20 DAI before upgrade: GasUsed[207302]
L2CustomERC20Gateway.withdrawERC20 DAI before upgrade: GasUsed[181053]
L2GatewayRouter.withdrawERC20 DAI before upgrade: GasUsed[204029]
      ✔ should succeed on L2CustomERC20Gateway (22519ms)
L2USDCGateway.withdrawERC20 USDC before upgrade: GasUsed[212736]
L2GatewayRouter.withdrawERC20 USDC before upgrade: GasUsed[229555]
L2USDCGateway.withdrawERC20 USDC before upgrade: GasUsed[203373]
L2GatewayRouter.withdrawERC20 USDC before upgrade: GasUsed[226282]
      ✔ should succeed on L2USDCGateway (31009ms)
```

File: contracts/docs/apis/L1ScrollMessenger.md [hunk 3]
```diff
@@ -611,3 +611,17 @@ Emitted when the maximum number of times each message can be replayed is updated
...
+
+### ErrorZeroAddress
+
+```solidity
+error ErrorZeroAddress()
+```
+
+
+
+*Thrown when the given address if `address(0)`.*
```
Comment (Thegaram) [inline contracts/docs/apis/L1ScrollMessenger.md]:
```suggestion
*Thrown when the given address is `address(0)`.*
```

File: contracts/docs/apis/L1ERC1155Gateway.md [hunk 4]
```diff
@@ -550,3 +550,50 @@ Emitted when token mapping for ERC1155 token is updated.
...
+
+### ErrorCallerIsNotCounterpartGateway
+
+```solidity
+error ErrorCallerIsNotCounterpartGateway()
+```
+
+
+
+*Thrown when the cross chain sender is not counterpart gateway contract.*
```
Comment (Thegaram) [inline contracts/docs/apis/L1ERC1155Gateway.md]:
```suggestion
*Thrown when the cross chain sender is not the counterpart gateway contract.*
```

File: contracts/integration-test/GasOptimizationUpgrade.spec.ts [hunk 5]
```diff
@@ -0,0 +1,588 @@
...
+        "L1GatewayRouter.depositERC20 WETH before upgrade"
+      );
+
+      // do upgrade
+      await upgradeL1(L1_GATEWAY, impl.address);
+
+      // after upgrade
+      await showGasUsage(
+        await gateway["depositERC20(address,uint256,uint256)"](L1_WETH, amountIn, 1e6, { value: fee }),
+        "L1WETHGateway.depositERC20 WETH before upgrade"
```
Comment (Thegaram) [inline contracts/integration-test/GasOptimizationUpgrade.spec.ts]:
```suggestion
        "L1WETHGateway.depositERC20 WETH after upgrade"
```

File: contracts/scripts/foundry/InitializeL1BridgeContracts.s.sol [hunk 6]
```diff
@@ -55,13 +73,18 @@ contract InitializeL1BridgeContracts is Script {
...
+        ProxyAdmin proxyAdmin = ProxyAdmin(L1_PROXY_ADMIN_ADDR);
+
         vm.startBroadcast(L1_DEPLOYER_PRIVATE_KEY);
 
         // initialize ScrollChain
-        ScrollChain(L1_SCROLL_CHAIN_PROXY_ADDR).initialize(
-            L1_MESSAGE_QUEUE_PROXY_ADDR,
-            L1_MULTIPLE_VERSION_ROLLUP_VERIFIER_ADDR,
-            MAX_TX_IN_CHUNK
+        proxyAdmin.upgradeAndCall(
```
Comment (Thegaram) [inline contracts/scripts/foundry/InitializeL1BridgeContracts.s.sol:82]: Is the idea here that we can now use these scripts to roll out this PR? (Deploy new contracts, then call `upgradeAndCall` on each proxy)

Although in practice, at least on mainnet, we cannot use this script since we'll need to go through the multisig.

File: contracts/src/L1/gateways/L1CustomERC20Gateway.sol [hunk 7]
```diff
@@ -54,8 +67,6 @@ contract L1CustomERC20Gateway is L1ERC20Gateway {
         address _router,
         address _messenger
     ) external initializer {
-        require(_router != address(0), "zero router address");
```
Comment (Thegaram) [inline contracts/src/L1/gateways/L1CustomERC20Gateway.sol:57]: Is it necessary to remove it here? (Just thinking in terms of minimizing the diff.)

File: contracts/src/L1/gateways/usdc/L1USDCGateway.sol [hunk 8]
```diff
@@ -49,23 +49,40 @@ contract L1USDCGateway is L1ERC20Gateway, IUSDCBurnableSourceBridge {
...
+    /// @param _router The address of `L1GatewayRouter` contract.
+    /// @param _messenger The address of `L1ScrollMessenger` contract.
+    constructor(
+        address _l1USDC,
+        address _l2USDC,
+        address _counterpart,
+        address _router,
+        address _messenger
+    ) ScrollGatewayBase(_counterpart, _router, _messenger) {
+        if (_router == address(0)) revert ErrorZeroAddress();
```
Comment (Thegaram) [inline contracts/src/L1/gateways/usdc/L1USDCGateway.sol]: Let' check the 2 token addresses too?

File: contracts/src/L1/rollup/IL1MessageQueueWithOracle.sol [hunk 9]
```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.16;
+
+import {IL1MessageQueue} from "./IL1MessageQueue.sol";
+
+interface IL1MessageQueueWithOracle is IL1MessageQueue {
```
Comment (Thegaram) [inline contracts/src/L1/rollup/IL1MessageQueueWithOracle.sol]:
```suggestion
interface IL1MessageQueueWithGasPriceOracle is IL1MessageQueue {
```

File: contracts/src/L1/rollup/IL1MessageQueueWithOracle.sol [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+    /// @notice Emitted when owner updates whitelist contract.
+    /// @param _oldWhitelist The address of old whitelist contract.
+    /// @param _newWhitelist The address of new whitelist contract.
+    event UpdateWhitelist(address _oldWhitelist, address _newWhitelist);
+
+    /// @notice Emitted when current l2 base fee is updated.
+    /// @param oldL2BaseFee The original l2 base fee before update.
+    /// @param newL2BaseFee The current l2 base fee updated.
+    event L2BaseFeeUpdated(uint256 oldL2BaseFee, uint256 newL2BaseFee);
```
Comment (Thegaram) [inline contracts/src/L1/rollup/IL1MessageQueueWithOracle.sol]: The name `UpdateL2BaseFee` would be more consistent (see also below).

File: contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+     *************/
+
+    /// @notice The latest known l2 base fee.
+    uint256 public l2BaseFee;
+
+    /// @notice The address of whitelist contract.
+    address public whitelist;
+
+    /// @notice The intrinsic params for transaction.
+    IntrinsicParams public intrinsicParams;
```
Comment (Thegaram) [inline contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol]: I feel like this will be updated very rarely, if ever, but it's read for each message, should we make it immutable?

File: contracts/src/L1/rollup/ScrollChain.sol [hunk 12]
```diff
@@ -98,24 +99,45 @@ contract ScrollChain is OwnableUpgradeable, PausableUpgradeable, IScrollChain {
...
+    ///
+    /// @param _chainId The chain id of L2.
+    /// @param _messageQueue The address of `L1MessageQueue` contract.
+    /// @param _verifier The address of zkevm verifier contract.
+    constructor(
+        uint64 _chainId,
+        address _messageQueue,
+        address _verifier
+    ) {
+        if (_verifier == address(0)) {
```
Comment (Thegaram) [inline contracts/src/L1/rollup/ScrollChain.sol]: Why not check `_messageQueue`?

File: contracts/src/libraries/IScrollMessenger.sol [hunk 13]
```diff
@@ -36,6 +31,13 @@ interface IScrollMessenger {
     /// @param messageHash The hash of the message.
     event FailedRelayedMessage(bytes32 indexed messageHash);
 
+    /**********
+     * Events *
```
Comment (Thegaram) [inline contracts/src/libraries/IScrollMessenger.sol]:
```suggestion
     * Errors *
```

File: contracts/scripts/foundry/InitializeL1BridgeContracts.s.sol (hunk 6 above)
Comment (zimpha) [reply contracts/scripts/foundry/InitializeL1BridgeContracts.s.sol:82]: Yeah, for fresh new deploy we can use there scripts.

File: contracts/src/L1/gateways/L1CustomERC20Gateway.sol (hunk 7 above)
Comment (zimpha) [reply contracts/src/L1/gateways/L1CustomERC20Gateway.sol:57]: removing it would reduce the codesize, it is ok to not change.

File: contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol (hunk 11 above)
Comment (zimpha) [reply contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol]: I think we can make it constant.

File: contracts/src/L1/gateways/L1ETHGateway.sol [hunk 14]
```diff
@@ -120,9 +133,12 @@ contract L1ETHGateway is ScrollGatewayBase, IL1ETHGateway, IMessageDropCallback
 
         // 1. Extract real sender if this call is from L1GatewayRouter.
         address _from = _msgSender();
+
+        /* comment out since router won't use this contract anymore
```
Comment (Thegaram) [inline contracts/src/L1/gateways/L1ETHGateway.sol:137]: Flow before: `L1GatewayRouter --> L1ETHGateway --> L1ScrollMessenger`.
Flow after: `L1GatewayRouter --> L1ScrollMessenger`. (No more `DepositETH` event emitted, same for `WithdrawETH` on L2.)

Will this break anything in `chain-monitor` or `bridge-history`? @colinlyguo @georgehao (I think it won't, just double checking.)

File: contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+pragma solidity =0.8.16;
+
+import {IWhitelist} from "../../libraries/common/IWhitelist.sol";
+import {IL1MessageQueue} from "./IL1MessageQueue.sol";
+import {IL1MessageQueueWithGasPriceOracle} from "./IL1MessageQueueWithGasPriceOracle.sol";
+import {IL2GasPriceOracle} from "./IL2GasPriceOracle.sol";
+
+import {L1MessageQueue} from "./L1MessageQueue.sol";
+
+contract L1MessageQueueWithGasPriceOracle is L1MessageQueue, IL1MessageQueueWithGasPriceOracle {
```
Comment (Thegaram) [inline contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol:12]: Note: we'll need to update `gas-oracle` configuration (contract address) right after upgrading this contract. cc @georgehao

File: contracts/src/L1/rollup/L1MessageQueue.sol [hunk 16]
```diff
@@ -20,37 +20,31 @@ import {AddressAliasHelper} from "../../libraries/common/AddressAliasHelper.sol"
 contract L1MessageQueue is OwnableUpgradeable, IL1MessageQueue {
```
Comment (Thegaram) [inline contracts/src/L1/rollup/L1MessageQueue.sol:20]:
```suggestion
abstract contract L1MessageQueue is OwnableUpgradeable, IL1MessageQueue {
```

Just suggesting this as a reminder that we plan not to deploy `L1MessageQueue` independently again.

If we do modify `L1MessageQueue` e.g. by adding a `reinitializer`, that will conflict with `initializeV2` in `L1MessageQueueWithGasPriceOracle`.

File: contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+    /// @inheritdoc IL1MessageQueue
+    function calculateIntrinsicGasFee(bytes calldata _calldata)
+        public
+        pure
+        override(IL1MessageQueue, L1MessageQueue)
+        returns (uint256)
+    {
+        // no way this can overflow `uint256`
+        unchecked {
+            return INTRINSIC_GAS_TX + _calldata.length * INTRINSIC_GAS_NONZERO_BYTE;
```
Comment (Thegaram) [inline contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol:77]: Should we add a reminder in a comment that this check will be insufficient if we ever cross-domain contract creation in the future?

File: contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+    /// @param _newL2BaseFee The new l2 base fee.
+    function setL2BaseFee(uint256 _newL2BaseFee) external {
+        if (!IWhitelist(whitelist).isSenderAllowed(_msgSender())) {
+            revert ErrorNotWhitelistedSender();
+        }
+
+        uint256 _oldL2BaseFee = l2BaseFee;
+        l2BaseFee = _newL2BaseFee;
+
+        emit UpdateL2BaseFee(_oldL2BaseFee, _newL2BaseFee);
```
Comment (Thegaram) [inline contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol:95]: Just realized that my suggestion to rename from `L2BaseFeeUpdated` to `UpdateL2BaseFee` might break any indexers that collect this event.

However, such indexers would need to be updated to a new contract address anyway, so I think it's acceptable.

File: contracts/src/L1/rollup/L2GasPriceOracle.sol [hunk 19]
```diff
@@ -40,7 +40,7 @@ contract L2GasPriceOracle is OwnableUpgradeable, IL2GasPriceOracle {
     uint256 public l2BaseFee;
 
     /// @notice The address of whitelist contract.
-    IWhitelist public whitelist;
```
Comment (Thegaram) [inline contracts/src/L1/rollup/L2GasPriceOracle.sol:43]: `L2GasPriceOracle` is no longer used, right? Why did you change these few lines?

File: contracts/integration-test/ScrollChain.spec.ts [hunk 20]
```diff
@@ -12,24 +12,31 @@ describe("ScrollChain", async () => {
...
 
     const ScrollChain = await ethers.getContractFactory("ScrollChain", deployer);
-    const chainImpl = await ScrollChain.deploy(0);
+    const chainImpl = await ScrollChain.deploy(0, queueProxy.address, deployer.address);
     await chainImpl.deployed();
-    const chainProxy = await TransparentUpgradeableProxy.deploy(chainImpl.address, admin.address, "0x");
-    await chainProxy.deployed();
+    await admin.upgrade(chainProxy.address, chainImpl.address);
+
+    queue = await ethers.getContractAt("L1MessageQueue", queueProxy.address, deployer);
```
Comment (Thegaram) [inline contracts/integration-test/ScrollChain.spec.ts:39]: Why not change this test to use `L1MessageQueueWithGasPriceOracle`?

File: contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol (hunk 15 above)
Comment (Thegaram) [inline contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol:12]: Is it possible to change the visibility of `gasOracle` from `public` (inherited from `L1MessageQueue`) to `internal`?

File: contracts/scripts/foundry/DeployL1BridgeContracts.s.sol [hunk 21]
```diff
@@ -85,26 +105,28 @@ contract DeployL1BridgeContracts is Script {
...
-        logAddress("L1_SCROLL_CHAIN_PROXY_ADDR", address(proxy));
     }
 
     function deployL1MessageQueue() internal {
-        L1MessageQueue impl = new L1MessageQueue();
-        TransparentUpgradeableProxy proxy = new TransparentUpgradeableProxy(
-            address(impl),
-            address(proxyAdmin),
-            new bytes(0)
+        L1MessageQueue impl = new L1MessageQueue(
```
Comment (Thegaram) [inline contracts/scripts/foundry/DeployL1BridgeContracts.s.sol]: Where is `L1MessageQueueWithGasPriceOracle` deployed?

File: contracts/scripts/foundry/DeployL1BridgeContracts.s.sol [hunk 22]
```diff
@@ -118,112 +140,87 @@ contract DeployL1BridgeContracts is Script {
...
-
-        logAddress("L1_STANDARD_ERC20_GATEWAY_IMPLEMENTATION_ADDR", address(impl));
-        logAddress("L1_STANDARD_ERC20_GATEWAY_PROXY_ADDR", address(proxy));
-    }
-
-    function deployL1ETHGateway() internal {
-        L1ETHGateway impl = new L1ETHGateway();
+    function deployL1GatewayRouter() internal {
+        L1GatewayRouter impl = new L1GatewayRouter(L1_SCROLL_MESSENGER_PROXY_ADDR);
         TransparentUpgradeableProxy proxy = new TransparentUpgradeableProxy(
```
Comment (Thegaram) [inline contracts/scripts/foundry/DeployL1BridgeContracts.s.sol:145]: Should we make it all consistent and deploy the proxy in the other script?

File: contracts/scripts/foundry/InitializeL1BridgeContracts.s.sol [hunk 23]
```diff
@@ -79,20 +102,34 @@ contract InitializeL1BridgeContracts is Script {
...
+            abi.encodeCall(
+                L1MessageQueue.initialize,
+                (
+                    L1_SCROLL_MESSENGER_PROXY_ADDR,
+                    L1_SCROLL_CHAIN_PROXY_ADDR,
+                    L1_ENFORCED_TX_GATEWAY_PROXY_ADDR,
+                    L2_GAS_PRICE_ORACLE_PROXY_ADDR,
+                    MAX_L1_MESSAGE_GAS_LIMIT
+                )
+            )
```
Comment (Thegaram) [inline contracts/scripts/foundry/InitializeL1BridgeContracts.s.sol:118]: Upgrade again and call `initializeV2` here?

File: contracts/src/test/L1CustomERC20Gateway.t.sol [hunk 24]
```diff
@@ -46,17 +47,18 @@ contract L1CustomERC20GatewayTest is L1GatewayTestBase {
...
         l1Token = new MockERC20("Mock L1", "ML1", 18);
         l2Token = new MockERC20("Mock L2", "ML2", 18);
 
-        // Deploy L1 contracts
-        gateway = _deployGateway();
-        router = L1GatewayRouter(address(new ERC1967Proxy(address(new L1GatewayRouter()), new bytes(0))));
-
         // Deploy L2 contracts
-        counterpartGateway = new L2CustomERC20Gateway();
+        counterpartGateway = new L2CustomERC20Gateway(address(1), address(1), address(1));
```
Comment (Thegaram) [inline contracts/src/test/L1CustomERC20Gateway.t.sol:57]: Is it okay to not set these properly?

File: contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+    /// @param _scrollChain The address of `ScrollChain` contract.
+    /// @param _enforcedTxGateway The address of `EnforcedTxGateway` contract.
+    constructor(
+        address _messenger,
+        address _scrollChain,
+        address _enforcedTxGateway
+    ) L1MessageQueue(_messenger, _scrollChain, _enforcedTxGateway) {}
+
+    /// @notice Initialize the storage of L1MessageQueueWithGasPriceOracle.
+    function initializeV2() external reinitializer(2) {
```
Comment (Thegaram) [inline contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol:49]: Should we ensure that the functions below can only be called once `initializeV2` has been called?

**Review by Thegaram — DISMISSED** [review]

File: contracts/src/L1/gateways/L1ETHGateway.sol (hunk 14 above)
Comment (georgehao) [reply contracts/src/L1/gateways/L1ETHGateway.sol:137]: yeah. chain monitor v2 & bridge history v2 all don't need any changing.

File: contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol (hunk 17 above)
Comment (zimpha) [reply contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol:77]: if we enable cross-domain contract creation, there should be some flag indicates it is contract creation. So, this will need another different function. I think no extra comments is needed.

File: contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol (hunk 18 above)
Comment (zimpha) [reply contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol:95]: agreed

File: contracts/src/L1/rollup/L2GasPriceOracle.sol (hunk 19 above)
Comment (zimpha) [reply contracts/src/L1/rollup/L2GasPriceOracle.sol:43]: `L1MessageQueueWithGasPriceOracle` will use the `IL2GasPriceOracle` interface, changing this to make sure file compile.

File: contracts/integration-test/ScrollChain.spec.ts (hunk 20 above)
Comment (zimpha) [reply contracts/integration-test/ScrollChain.spec.ts:39]: The scroll chain unit tests don't need functions provided by `L1MessageQueueWithGasPriceOracle`, using `L1MessageQueue` is enough.

File: contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol (hunk 15 above)
Comment (zimpha) [reply contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol:12]: It is possible, but it will change the existing contract abi. I suggest not to do this.

File: contracts/src/test/L1CustomERC20Gateway.t.sol (hunk 24 above)
Comment (zimpha) [reply contracts/src/test/L1CustomERC20Gateway.t.sol:57]: `counterpartGateway` is just used as a placeholder contract, it is ok to not set these properly.

File: contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol (hunk 25 above)
Comment (zimpha) [reply contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol:49]: not necessary, will introduce extra gas costs.

File: contracts/scripts/foundry/DeployL1BridgeContracts.s.sol (hunk 22 above)
Comment (zimpha) [reply contracts/scripts/foundry/DeployL1BridgeContracts.s.sol:145]: yes, we can. But since this script is very likely won't be used, we can do these changes in another PR.

File: contracts/src/L1/rollup/L1MessageQueue.sol (hunk 16 above)
Comment (zimpha) [reply contracts/src/L1/rollup/L1MessageQueue.sol:20]: Well, normally this won't happen. We can keep it `non-abstract`. And it is also easy to write unit tests.

File: contracts/src/L1/gateways/L1ETHGateway.sol (hunk 14 above)
Comment (colinlyguo) [reply contracts/src/L1/gateways/L1ETHGateway.sol:137]: if the gateway contracts are upgraded, do the chain-monitor and bridge-history need to monitor an "array" of contracts? or the proxy addresses would remain the same, thus the addresses in the config file would not change?

**Review by Thegaram — DISMISSED** [review]

File: contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol [hunk 26]
```diff
@@ -0,0 +1,110 @@
...
+
+    /*************
+     * Variables *
+     *************/
+
+    /// @notice The latest known l2 base fee.
+    uint256 public l2BaseFee;
+
+    /// @notice The address of whitelist contract.
+    address public whitelist;
```
Comment (icemelon) [inline contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol]: should not call it whitelist since it's only a single owner?

File: contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol [hunk 27]
```diff
@@ -0,0 +1,110 @@
...
+        external
+        view
+        override(IL1MessageQueue, L1MessageQueue)
+        returns (uint256)
+    {
+        return _gasLimit * l2BaseFee;
+    }
+
+    /// @inheritdoc IL1MessageQueue
+    function calculateIntrinsicGasFee(bytes calldata _calldata)
```
Comment (icemelon) [inline contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol:69]: Realize that this function name is confusing since it only calculates the intrinsic gas, not the fee. Should we remove the "Fee" from the name?

Comment (zimpha) [reply contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol:69]: It is not good to change the public abi, thirdparty devs may already use this.

File: contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol (hunk 26 above)
Comment (zimpha) [reply contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol]: yeah, maybe better to call it `whitelistChecker`.

**Review by icemelon — APPROVED** [review]

File: contracts/src/L1/gateways/L1ETHGateway.sol (hunk 14 above)
Comment (Thegaram) [reply contracts/src/L1/gateways/L1ETHGateway.sol:137]: The proxy addresses will remain the same, no need to change the config (except for gas oracle, which is merged into L1MessageQueue).

**Review by Thegaram — APPROVED** [review]

File: contracts/src/L1/gateways/L1ETHGateway.sol (hunk 14 above)
Comment (colinlyguo) [reply contracts/src/L1/gateways/L1ETHGateway.sol:137]: Got it. Then this PR would not introduce incompatibility to the monitoring services (chain-monitor, message fetcher of bridge-history) and rollup-relayer, only need to change `gas-oracle` sender's target address.
